### ISSUE TITLE
KeyError: 'sum'

### ISSUE THREAD

**Issue by mdozmorov** [dejunlin/hicrep#9]
Hi. I've installed `hicrep` into Python 3.8 venv. The command line version breaks with the `KeyError: 'sum'` error:
````
hicrep --binSize 100000 --h 10 --dBPMax 5000000 03.mcool 04.mcool 03_04.txt
Traceback (most recent call last):
  File "/home/mdozmorov/miniconda3/bin/hicrep", line 8, in <module>
    sys.exit(main())
  File "/home/mdozmorov/miniconda3/lib/python3.8/site-packages/hicrep/__init__.py", line 83, in main
    scc = hicrepSCC(cool1, cool2, h, dBPMax, bDownSample)
  File "/home/mdozmorov/miniconda3/lib/python3.8/site-packages/hicrep/hicrep.py", line 146, in hicrepSCC
    n1 = cool1.info['sum']
KeyError: 'sum'
````

Trying from Python results in a different error:
````
from hicrep.utils import readMcool
fmcool1 = "/home/sequencing/j_03.mcool"
fmcool2 = "/home/sequencing/04.mcool"
binSize = 100000
cool1, binSize1 = readMcool(fmcool1, binSize)
cool2, binSize2 = readMcool(fmcool2, binSize)
h = 10
dBPMax = 500000
bDownSample = False
scc = hicrepSCC(cool1, cool2, h, dBPMax, bDownSample)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
NameError: name 'hicrepSCC' is not defined
````
Any advice?
Thanks,
Mikhail

**Comment by dejunlin** [comment]
For the second error, you need to import the hicrepSCC function. See the
instruction on the front page on how to do this.

For the first error, all Cooler files I have dealt with have a python
dictionary cool.info of various metadata and "sum" is one of its field
indicating the total read counts. Your Cooler file is a bit unusual and
obviously lacks that "sum" field. Can you post what's in your Cooler file's
info field? How did you generate that Cooler file and what version of
software did you use?

**Comment by mdozmorov** [comment]
I used https://github.com/4dn-dcic/hic2cool, to convert `.hic` to `.mcool`. It worked, but indeed, calling `cooler info` on them results in a long error ending:
````
KeyError: "Unable to open object (object 'chroms' doesn't exist)"
KeyError: "No cooler found at: 01.mcool. Coolers found in ['/resolutions/5000', '/resolutions/10000', '/resolutions/25000', '/resolutions/50000', '/resolutions/100000', '/resolutions/250000', '/resolutions/500000', '/resolutions/1000000', '/resolutions/2500000']. Use '::' to specify a group path"
````
I'm fairly new to `.cool` format, so not sure what to do if the `hic2cool` tool doesn't work.

Most surprisingly, I used `hicrep` about 4 months ago on that same files, all worked. Now, `hicrep` has been updated and requires `--h` and `--dBPMax` arguments. And throws an error.

**Comment by dejunlin** [comment]
I have used hic2cool a lot and never had such problems. Anyway, at least
you've got a error to warn you of the situation. If I were you, I'd worry
if any step upstream had made a silent error somewhere, which could have
corrupted your data in some unknown way without telling you so. I can think
of a workaround for this error but you need to make sure your input data
conforms to standard in order to get expected result.

BTW, for the --h and --dBPMax options, you are encouraged to set those
options according to the application even in the earlier version and the
reason for this has been discussed in the HiCRep paper

HiCRep: assessing the reproducibility of Hi-C data using a stratum-adjusted
correlation coefficient. Tao Yang Feipeng Zhang Galip Gürkan Yardımcı Fan
Song Ross C. Hardison William Stafford Noble Feng Yue and Qunhua Li, Genome
Res. 2017 Nov;27(11):1939-1949. doi: 10.1101/gr.220640.117.

please read the paper before you use the tool for scientific work.

**Comment by dejunlin** [comment]
@mdozmorov Can you check if your input data have at least these `cool.info` fields: `cool.info['bin-size']`, `cool.info['nbins']`, `cool.info['nchroms']` and `cool.info['nnz']`? If they are available, I can probably work around the `sum` thing; otherwise, you need to make sure these fields are in the data in order for the tool to work

**Comment by mdozmorov** [comment]
Those fields are there.
````
>>> cool1.info['bin-size']
100000
>>> cool1.info['nbins']
30895
>>> cool1.info['nchroms']
25
>>> cool1.info['nnz']
40900050
````
And the overall output of `cool.info` resembles the header of a `.hic` file.

Re: `hic2cool` - I intentionally reran it to generate a "fresh" `.mcool` file. Same error. For clarity, I've installed `hic2cool` to a new `conda` environment and currently regenerating the `.mcool` file. Not sure it this is the problem thoubh.

**Comment by dejunlin** [comment]
@mdozmorov I provide an alternative path of computing the 'sum' metadata field in https://github.com/dejunlin/hicrep/commit/66b3fa3670632d60d1ce0573f2f37aaf763279f5, which will be released in version 0.2.4 when the TravisCI bot finishes the testing and deployment into pypi. You are welcome to try the patch out directly from the source code repo if you are eager to test it out because it might take the bot a few days to flush everything out. 

I think this should fix the problem and I'm closing this issue for now but feel free to comment and propose suggestion if this doesn't work for you.

**Comment by mdozmorov** [comment]
Thanks, @dejunlin - tested, it works. Also, tested the new function and confirm that, when I use `cool1.info('sum')`, I get an error:
````
>>> cool1.info('sum')
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: 'dict' object is not callable
````
When using `coolerInfo` on the same file, it works:
````
>>> from hicrep.utils import coolerInfo
>>> coolerInfo(cool1, 'sum')
119971895
````
I tested both the files converted with the latest `hic2cool`, and those converted 4 months ago. None have the `.info('sum')` field. Given HiCrep works now, I guess we leave it at that. Thanks again!
